### ISSUE TITLE
RuntimeWarning encountered for unary_union operations.

### ISSUE THREAD

**Issue by Aniwax** [pygeos/pygeos#357]
I have a set of polygons:

```
'POLYGON ((-0.961 0.109, -0.16 -0.98, -0.782 -0.319, -0.961 0.109))',
 'POLYGON ((-0.782 -0.319, -0.622 0.661, -0.961 0.109, -0.782 -0.319))',
 'POLYGON ((-0.179 0.428, -0.961 0.109, -0.622 0.661, -0.179 0.428))',
 'POLYGON ((0.179 -0.428, -0.16 -0.98, 0.622 -0.661, 0.179 -0.428))',
 'POLYGON ((0.179 -0.428, -0.782 -0.319, -0.16 -0.98, 0.179 -0.428))',
 'POLYGON ((0.179 -0.428, -0.622 0.661, -0.782 -0.319, 0.179 -0.428))',
 'POLYGON ((0.782 0.319, 0.622 -0.661, -0.179 0.428, 0.782 0.319))',
 'POLYGON ((0.16 0.98, -0.179 0.428, -0.622 0.661, 0.16 0.98))',
 'POLYGON ((0.16 0.98, 0.782 0.319, -0.179 0.428, 0.16 0.98))',
 'POLYGON ((-0.339 -0.552, -0.179 0.428, 0.622 -0.661, -0.339 -0.552))',
 'POLYGON ((-0.339 -0.552, -0.16 -0.98, -0.961 0.109, -0.339 -0.552))'
```

when performing `union_all()` on them, 2 `RuntimeWarning` are generated:

```
polygons = pygeos.from_wkt(
      ['POLYGON ((-0.961 0.109, -0.16 -0.98, -0.782 -0.319, -0.961 0.109))',
       'POLYGON ((-0.782 -0.319, -0.622 0.661, -0.961 0.109, -0.782 -0.319))',
       'POLYGON ((-0.179 0.428, -0.961 0.109, -0.622 0.661, -0.179 0.428))',
       'POLYGON ((0.179 -0.428, -0.16 -0.98, 0.622 -0.661, 0.179 -0.428))',
       'POLYGON ((0.179 -0.428, -0.782 -0.319, -0.16 -0.98, 0.179 -0.428))',
       'POLYGON ((0.179 -0.428, -0.622 0.661, -0.782 -0.319, 0.179 -0.428))',
       'POLYGON ((0.782 0.319, 0.622 -0.661, -0.179 0.428, 0.782 0.319))',
       'POLYGON ((0.16 0.98, -0.179 0.428, -0.622 0.661, 0.16 0.98))',
       'POLYGON ((0.16 0.98, 0.782 0.319, -0.179 0.428, 0.16 0.98))',
       'POLYGON ((-0.339 -0.552, -0.179 0.428, 0.622 -0.661, -0.339 -0.552))',
       'POLYGON ((-0.339 -0.552, -0.16 -0.98, -0.961 0.109, -0.339 -0.552))'
      ])      
pygeos.union_all(polygons )
```

```
C:\ibrus-env\lib\site-packages\pygeos\set_operations.py:388: RuntimeWarning: divide by zero encountered in unary_union
  result = lib.unary_union(collections, **kwargs)
C:\ibrus-env\lib\site-packages\pygeos\set_operations.py:388: RuntimeWarning: invalid value encountered in unary_union
  result = lib.unary_union(collections, **kwargs)
```

I don't know what exactly triggered this warning and if it affects the calculation result of union operations.

**Comment by jorisvandenbossche** [comment]
It might also depend on the GEOS version (i.e. I can reproduce it in an environment with GEOS 3.9.1, but not in an environment with GEOS 3.8.1 or GEOS master; although some other things also vary across those environments ...)

**Comment by Aniwax** [comment]
I produced this with pygeos 0.10. Not sure which version of GEOS is it running on. How can I check it?

**Comment by jorisvandenbossche** [comment]
You can check `pygeos.geos_version`

**Comment by Aniwax** [comment]
GEOS version 3.9.1 as well

**Comment by mwtoews** [comment]
For the record, shapely seems to be OK with this, shown here:
```python
from shapely import wkt, ops

polygons_str = (['POLYGON ((-0.961 0.109, -0.16 -0.98, -0.782 -0.319, -0.961 0.109))',
    'POLYGON ((-0.782 -0.319, -0.622 0.661, -0.961 0.109, -0.782 -0.319))',
    'POLYGON ((-0.179 0.428, -0.961 0.109, -0.622 0.661, -0.179 0.428))',
    'POLYGON ((0.179 -0.428, -0.16 -0.98, 0.622 -0.661, 0.179 -0.428))',
    'POLYGON ((0.179 -0.428, -0.782 -0.319, -0.16 -0.98, 0.179 -0.428))',
    'POLYGON ((0.179 -0.428, -0.622 0.661, -0.782 -0.319, 0.179 -0.428))',
    'POLYGON ((0.782 0.319, 0.622 -0.661, -0.179 0.428, 0.782 0.319))',
    'POLYGON ((0.16 0.98, -0.179 0.428, -0.622 0.661, 0.16 0.98))',
    'POLYGON ((0.16 0.98, 0.782 0.319, -0.179 0.428, 0.16 0.98))',
    'POLYGON ((-0.339 -0.552, -0.179 0.428, 0.622 -0.661, -0.339 -0.552))',
    'POLYGON ((-0.339 -0.552, -0.16 -0.98, -0.961 0.109, -0.339 -0.552))'
])
polygons = [wkt.loads(t) for t in polygons_str]
print(ops.unary_union(polygons).wkt)
```
with GEOS 3.9.1:
```
POLYGON ((-0.961 0.109, -0.622 0.661, 0.16 0.98, 0.782 0.319, 0.622 -0.661, -0.16 -0.98, -0.782 -0.319, -0.961 0.109), (-0.2283390880090073 0.1257980859448302, -0.179 0.428, -0.3879308609659142 0.3427711705266923, -0.2283390880090073 0.1257980859448302))
```
with GEOS 3.8.0:
```
POLYGON ((0.622 -0.661, -0.16 -0.98, -0.782 -0.319, -0.961 0.109, -0.622 0.661, 0.16 0.98, 0.782 0.319, 0.622 -0.661), (-0.2283390880090073 0.1257980859448302, -0.179 0.428, -0.3879308609659142 0.3427711705266923, -0.2283390880090073 0.1257980859448302))
```
which is the equal to 3.9.1 when normalized.

**Comment by Aniwax** [comment]
I can 

> For the record, shapely seems to be OK with this, shown here:
> 
> ```python
> from shapely import wkt, ops
> 
> polygons_str = (['POLYGON ((-0.961 0.109, -0.16 -0.98, -0.782 -0.319, -0.961 0.109))',
>     'POLYGON ((-0.782 -0.319, -0.622 0.661, -0.961 0.109, -0.782 -0.319))',
>     'POLYGON ((-0.179 0.428, -0.961 0.109, -0.622 0.661, -0.179 0.428))',
>     'POLYGON ((0.179 -0.428, -0.16 -0.98, 0.622 -0.661, 0.179 -0.428))',
>     'POLYGON ((0.179 -0.428, -0.782 -0.319, -0.16 -0.98, 0.179 -0.428))',
>     'POLYGON ((0.179 -0.428, -0.622 0.661, -0.782 -0.319, 0.179 -0.428))',
>     'POLYGON ((0.782 0.319, 0.622 -0.661, -0.179 0.428, 0.782 0.319))',
>     'POLYGON ((0.16 0.98, -0.179 0.428, -0.622 0.661, 0.16 0.98))',
>     'POLYGON ((0.16 0.98, 0.782 0.319, -0.179 0.428, 0.16 0.98))',
>     'POLYGON ((-0.339 -0.552, -0.179 0.428, 0.622 -0.661, -0.339 -0.552))',
>     'POLYGON ((-0.339 -0.552, -0.16 -0.98, -0.961 0.109, -0.339 -0.552))'
> ])
> polygons = [wkt.loads(t) for t in polygons_str]
> print(ops.unary_union(polygons).wkt)
> ```
> 
> with GEOS 3.9.1:
> 
> ```
> POLYGON ((-0.961 0.109, -0.622 0.661, 0.16 0.98, 0.782 0.319, 0.622 -0.661, -0.16 -0.98, -0.782 -0.319, -0.961 0.109), (-0.2283390880090073 0.1257980859448302, -0.179 0.428, -0.3879308609659142 0.3427711705266923, -0.2283390880090073 0.1257980859448302))
> ```
> 
> with GEOS 3.8.0:
> 
> ```
> POLYGON ((0.622 -0.661, -0.16 -0.98, -0.782 -0.319, -0.961 0.109, -0.622 0.661, 0.16 0.98, 0.782 0.319, 0.622 -0.661), (-0.2283390880090073 0.1257980859448302, -0.179 0.428, -0.3879308609659142 0.3427711705266923, -0.2283390880090073 0.1257980859448302))
> ```
> 
> which is the equal to 3.9.1 when normalized.

I can confirm this. I didn't have this warning when I was purely using` Shapely`. Now my team is completely switching to `pygeos`, for performance reason. However, it's too cumbersome to manage `Shapely` and `pygeos` in the same environment for the user of our package. So it would be great to be able to rely on `pygeos` completely.

**Comment by martinfleis** [comment]
I just want to note that pygeos returns the correct result, the same as shapely. It just emits those warnings along the way.

**Comment by caspervdw** [comment]
Looking at our test suite it seems that many of these warnings were solved in GEOS 3.10. 

Still, there are some left. Most of them are emitted when there are NaN coordinates in an input geometry, which is acceptable I think. Ideally the warnings should be filtered out in those specific test cases. So I leave this issue open to address that.
